File: app/lib/domain/plan_schema.dart
import 'dart:convert';

import 'models/plan.dart' show Plan;

class PlanValidationResult {
  final Plan? plan;
  final String? error;
  const PlanValidationResult.success(this.plan) : error = null;
  const PlanValidationResult.failure(this.error) : plan = null;
}

bool looksLikePlanJson(String text) {
  final trimmed = text.trim();
  if (!trimmed.startsWith('{') || !trimmed.endsWith('}')) return false;
  try {
    jsonDecode(trimmed);
    return true;
  } catch (_) {
    return false;
  }
}

PlanValidationResult validatePlanJson(String rawJson) {
  final trimmed = rawJson.trim();
  Map<String, dynamic>? json;
  try {
    final decoded = jsonDecode(trimmed);
    if (decoded is! Map<String, dynamic>) {
      return const PlanValidationResult.failure('JSON 顶层必须是对象');
    }
    json = decoded;
  } catch (e) {
    return PlanValidationResult.failure('JSON 解析失败: $e');
  }

  final required = ['weeks', 'weekly_structure', 'goal_summary', 'training_days', 'daily_meals'];
  for (final key in required) {
    if (!json.containsKey(key)) {
      return PlanValidationResult.failure('缺少必需字段: $key');
    }
  }

  if (json['weeks'] is! int || (json['weeks'] as int) <= 0) {
    return const PlanValidationResult.failure('weeks 必须是正整数');
  }
  if (json['training_days'] is! List) {
    return const PlanValidationResult.failure('training_days 必须是数组');
  }
  if (json['daily_meals'] is! List) {
    return const PlanValidationResult.failure('daily_meals 必须是数组');
  }

  try {
    final startDate = json['startDate'] is String
        ? DateTime.parse(json['startDate'] as String)
        : DateTime.now();
    final planWithDate = {...json, 'startDate': startDate.toIso8601String()};
    final plan = Plan.fromJson(planWithDate);
    return PlanValidationResult.success(plan);
  } catch (e) {
    return PlanValidationResult.failure('Plan 字段类型不匹配: $e');
  }
}

File: app/android/app/src/main/java/com/gymcoach/gym_coach/domain/plan/PlanSchema.kt
package com.gymcoach.gym_coach.domain.plan

import com.gymcoach.gym_coach.domain.model.Plan
import kotlinx.serialization.SerializationException
import kotlinx.serialization.json.Json

/**
 * Mirrors `lib/domain/plan_schema.dart`. `looksLikePlanJson` is a fast guard
 * used by the onboarding screen; `validatePlanJson` performs full validation.
 */
object PlanSchema {

    private val json = Json {
        ignoreUnknownKeys = true
        coerceInputValues = false
    }

    fun looksLikePlanJson(text: String): Boolean {
        val trimmed = text.trim()
        if (!trimmed.startsWith("{") || !trimmed.endsWith("}")) return false
        return runCatching { json.parseToJsonElement(trimmed) }.isSuccess
    }

    fun validatePlanJson(rawJson: String): PlanValidationResult {
        val trimmed = rawJson.trim()
        val parsed = runCatching { json.parseToJsonElement(trimmed) }.getOrNull()
            ?: return PlanValidationResult.Failure("not_json")
        val obj = parsed as? kotlinx.serialization.json.JsonObject
            ?: return PlanValidationResult.Failure("not_object")

        val required = listOf("weeks", "weekly_structure", "goal_summary", "training_days", "daily_meals")
        for (key in required) {
            if (!obj.containsKey(key)) return PlanValidationResult.Failure("missing_$key")
        }
        val weeks = (obj["weeks"] as? kotlinx.serialization.json.JsonPrimitive)?.content?.toIntOrNull()
            ?: return PlanValidationResult.Failure("bad_weeks")
        if (weeks <= 0) return PlanValidationResult.Failure("bad_weeks")
        if (obj["training_days"] !is kotlinx.serialization.json.JsonArray) return PlanValidationResult.Failure("bad_training_days")
        if (obj["daily_meals"]   !is kotlinx.serialization.json.JsonArray) return PlanValidationResult.Failure("bad_daily_meals")

        val withStart = if (obj.containsKey("start_date")) {
            trimmed
        } else {
            val now = java.time.Instant.now().toString()
            val mutable = obj.toMutableMap()
            mutable["start_date"] = kotlinx.serialization.json.JsonPrimitive(now)
            json.encodeToString(kotlinx.serialization.json.JsonObject.serializer(), kotlinx.serialization.json.JsonObject(mutable))
        }
        return try {
            val plan = json.decodeFromString(Plan.serializer(), withStart)
            PlanValidationResult.Success(plan)
        } catch (e: SerializationException) {
            PlanValidationResult.Failure("serialize:${e.message}")
        }
    }
}

sealed class PlanValidationResult {
    data class Success(val plan: Plan) : PlanValidationResult()
    data class Failure(val error: String) : PlanValidationResult()
}
